devices/src/virtio/queue.rs

```diff
@@ -142,65 +142,6 @@ impl<'a> DescriptorChain<'a> {
     }
 }
 
-/// Consuming iterator over all available descriptor chain heads in the queue.
-pub struct AvailIter<'a, 'b> {
-    mem: &'a GuestMemory,
-    desc_table: GuestAddress,
-    avail_ring: GuestAddress,
-    next_index: Wrapping<u16>,
-    last_index: Wrapping<u16>,
-    queue_size: u16,
-    next_avail: &'b mut Wrapping<u16>,
-}
-
-impl<'a, 'b> AvailIter<'a, 'b> {
-    pub fn new(mem: &'a GuestMemory, q_next_avail: &'b mut Wrapping<u16>) -> AvailIter<'a, 'b> {
-        AvailIter {
-            mem,
-            desc_table: GuestAddress(0),
-            avail_ring: GuestAddress(0),
-            next_index: Wrapping(0),
-            last_index: Wrapping(0),
-            queue_size: 0,
-            next_avail: q_next_avail,
-        }
-    }
-}
-
-impl<'a, 'b> Iterator for AvailIter<'a, 'b> {
```

Comment:
This means that `for pkt in queue { ... }` will no longer work. Is this really desirable?

Reply:
We were never actually using the `for desc in queue` pattern anywhere, because we couldn't run a process/mark-used cycle (since the iterator would hold a mutable reference to `queue`):

```rust
for desc in queue {
  process(desc);
  // This fails, because `queue` is already mutably borrowed via the iterator.
  queue.add_used(desc);
}
```

So now we do something like:

```rust
while Some(desc) = queue.pop() {
  process(desc);
  queue.add_used(desc);
}
```

Using an iterator over the avail queue would require this pattern, which I find kinda awkward:

```rust
let used = Vec::new();
for desc in queue {
  process(desc);
  used.pushed(desc);
}
for desc in used {
  queue.add_used(desc);
}
```

If we really need an iterator, for some use-case that I can't currently imagine, we can easily add one that uses `queue.pop()` to yield `next()`. The old iterator was unnecessarily complex, copying some data from the queue, while still holding a mutable reference to `queue.next_avail`.